gap/digraph.gi

```diff
@@ -1368,6 +1423,259 @@ function(filt, n, p)
+  
+    adjacencyList := [];
+  
+    for i in [1 .. n] do
+        Add(adjacencyList, []);
+    od;
+  
+    # Edge Case
+    if n = 1 then
+        if Float(Random([0 .. 99])/100) < Float(p) then
+            Add(adjacencyList[1], 1);
+            return Digraph(adjacencyList);
+        else
+            return Digraph(adjacencyList);
+        fi;
+    fi;
+  
+    vertices:= [1 .. n];
+    
+    # Starting from a random vertex, we create a Hamiltonian cycle
+    startVertex := Remove(vertices, Random(vertices));
+    hamiltonianCycle := [startVertex];
+    
+    # While there are remaining n-1 vertices to be added to the Hamiltonian cycle
+    for x in [1 .. n-1] do
+        # Create a random edge from the last vertex in the cycle, to random vertex not in the cycle
+        i := hamiltonianCycle[x];
+        j := Random(vertices);
+        Remove(vertices, Position(vertices, j));
+        Add(hamiltonianCycle, j);
+        Add(adjacencyList[i], j);
+    od;
+    
+    Add(adjacencyList[hamiltonianCycle[n]], startVertex);
+    
+    # Once we have created a Hamiltonian cycle, fill out the rest of the graph with random edges according to p
+    adjacencyList := DIGRAPHS_FillOutGraph(n, p, adjacencyList);
+    graph := Digraph(adjacencyList);
+    return graph;
+end);
+
+InstallMethod(RandomDigraphCons,
+"for IsEulerianDigraph, a positive integer, and a float",
+[IsEulerianDigraph, IsPosInt, IsFloat],
+function(filt, n, p)
+    local adjacencyList, vertices, startVertex, circuitVertex, continueCircuit,  verticesToConsider, connectedVertices, verticesToCheck, i, graph;
+
+    adjacencyList := [];
+    for i in [1 .. n] do
+        Add(adjacencyList, []);
+    od;
+
+    # Edge Case
+    if n = 1 then
+        if Float(Random([0 .. 99])/100) < Float(p) then
+            Add(adjacencyList[1], 1);
+            return Digraph(adjacencyList);
+        else
+            return Digraph(adjacencyList);
+        fi;
+    fi;
+
+    vertices:= [1 .. n];
+
+    # Starting from a random vertex, we create an Eulerian circuit
+    startVertex := Random(vertices);
+    circuitVertex := startVertex;
+    continueCircuit := true;
+
+    # This will keep track of the vertices left to consider when deciding whether to extend the cycle. For example,
+    # if we want to extend from the current circuitVertex, we can check - verticesToConsider[circuitVertex] -
+    # to see which vertices it hasn't previously considered.
+    verticesToConsider := [];
+    for i in [1 .. n] do
+        Add(verticesToConsider, [1 .. n]);
+    od;
+
+    # Eulerian graphs must be connected, so we store a list of connected vertices and check if a vertex isn't in the list
+    connectedVertices := [startVertex];
+
+    while continueCircuit do
+        while Length(verticesToConsider[circuitVertex]) > 0 do
+            # From the remaining vertices to check, select a random one to see if an edge will be added
+            i := Random(verticesToConsider[circuitVertex]);
```

Comment:
As above, better to select a random *position* rather than a random *element*.  Then we don't have to use `Position` below, which is expensive.